zone/client.cpp

```diff
@@ -10129,3 +10137,450 @@ void Client::SetAFK(uint8 afk_flag) {
+					{
+						//still potential to trigger this if you're knocked back off a
+						//HUGE fall that takes > 2.5 seconds
+						if (speed > 30.0f)
+						{
+							CheatDetected(MQWarpKnockBack, GetX(), GetY(), GetZ());
+						}
+					}
+					else if (!IsPortExempted())
+					{
+						if (!IsMQExemptedArea(zone->GetZoneID(), GetX(), GetY(), GetZ()))
+						{
+							if (speed > (runs * 2 * RuleR(Zone, MQWarpDetectionDistanceFactor)))
+							{
+								m_time_since_last_position_check = cur_time;
+								m_distance_since_last_position_check = 0.0f;
+								CheatDetected(MQWarp, GetX(), GetY(), GetZ());
+							}
+							else
+							{
+								CheatDetected(MQWarpLight, GetX(), GetY(), GetZ());
+							}
+						}
+					}
+				}
+			}
+		}
+		m_time_since_last_position_check = cur_time;
+		m_distance_since_last_position_check = 0.0f;
+	}
+	SetKnockBackExemption(v);
+}
+
+void Client::SetPortExemption(bool v)
+{
+	if (v == true)
+	{
+		uint32 cur_time = Timer::GetCurrentTime();
+		if ((cur_time - m_time_since_last_position_check) > 1000)
+		{
+			float speed = (m_distance_since_last_position_check * 100) / (float)(cur_time - m_time_since_last_position_check);
+			int runs = GetRunspeed();
+			if (speed > (runs * RuleR(Zone, MQWarpDetectionDistanceFactor)))
+			{
+				if (!GetGMSpeed() && (runs >= GetBaseRunspeed() || (speed > (GetBaseRunspeed() * RuleR(Zone, MQWarpDetectionDistanceFactor)))))
+				{
+					LogCheat(
+						"[SetPortExemption] Client [{}] moving too fast! Moved [{}] in [{}](s) speed [{}]",
+						GetCleanName(),
+						m_distance_since_last_position_check,
+						(cur_time - m_time_since_last_position_check),
+						speed
+					);
+					if (IsShadowStepExempted())
+					{
+						if (m_distance_since_last_position_check > 800)
+						{
+							CheatDetected(MQWarpShadowStep, GetX(), GetY(), GetZ());
+						}
+					}
+					else if (IsKnockBackExempted())
+					{
+						//still potential to trigger this if you're knocked back off a
+						//HUGE fall that takes > 2.5 seconds
+						if (speed > 30.0f)
+						{
+							CheatDetected(MQWarpKnockBack, GetX(), GetY(), GetZ());
+						}
+					}
+					else if (!IsPortExempted())
+					{
+						if (!IsMQExemptedArea(zone->GetZoneID(), GetX(), GetY(), GetZ()))
+						{
+							if (speed > (runs * 2 * RuleR(Zone, MQWarpDetectionDistanceFactor)))
+							{
+								m_time_since_last_position_check = cur_time;
+								m_distance_since_last_position_check = 0.0f;
+								CheatDetected(MQWarp, GetX(), GetY(), GetZ());
+							}
+							else
+							{
+								CheatDetected(MQWarpLight, GetX(), GetY(), GetZ());
+							}
+						}
+					}
+				}
+			}
+		}
+		m_time_since_last_position_check = cur_time;
+		m_distance_since_last_position_check = 0.0f;
+	}
+	SetPortExemption(v);
+}
+
+const bool Client::IsMQExemptedArea(uint32 zoneID, float x, float y, float z) const
+{
+	float max_dist = 90000;
+	switch (zoneID)
+	{
+	case 2:
+	{
+		float f_delta = (x - (-713.6));
```

Comment:
Ah, your variable names were fine on these, I meant putting the `f` after the floats
so
`float delta = (x - (-713.6));`
would become
`float f_delta = (x - (-713.6f));`

Reply:
ohhh okay, yeah, i can do that, thanks for the clarification lol